Fix #2711 & part of #5343: Workflow cancellation support in CI using concurrency

Rd4dev opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fix #2711 
Fixes part of #5343 

### Project
[PR 2.5 of Project 4.1]

### Changes Made
- The existing workflow canceller utilized the [cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) by syfle.
- Since GitHub introduced native [concurrency](https://docs.github.com/en/actions/using-jobs/using-concurrency) support on April 19, 2021, we can now utilize this built-in feature for better integration.
- This implementation leverages the [concurrency property](https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-concurrency-and-the-default-behavior) to manage workflows, using the configuration:
```
concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true
```
- This setup groups workflows by their name and reference, allowing GitHub to automatically cancel any in-progress jobs in the workflow when a new workflow run is triggered.
- This retires the existing Automatic workflow canceller (workflow_canceller.yml).
- This approach is now in place for the following workflows: 
  - Build Tests 
  - Static Checks 
  - Unit Tests (Robolectric - Gradle)
  - Unit Tests (Robolectric - Bazel)

### Triggered Cancellation
(PR Check Triggers) -> https://github.com/oppia/oppia-android/pull/5466/checks

### Screenshots of workflow cancelled
**Workflows Running:**

![Screenshot (1536)](https://github.com/user-attachments/assets/fa11a887-0bee-4d6c-8c58-fa2e723bde91)

**Workflows in previous run cancelling after latest commit**

![Screenshot (1538)](https://github.com/user-attachments/assets/14d96ff7-38c6-4054-a084-7712c34b9c29)

**Note:** It is observed that the Non-App Module runs take a bit of time to cancel but they eventually get cancelled.

![image](https://github.com/user-attachments/assets/516628ee-15a0-4a95-926c-ea925357bae2)

![image](https://github.com/user-attachments/assets/03a74685-c458-44a8-ada9-09faf45eeb93)

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Rd4dev committed:
Workflow cancellation with concurrency

Rd4dev committed:
Trigger #1 Triggering CI builds to check workflow cancellations with concurrency

Rd4dev committed:
Trigger #2 Triggering CI builds #1 worked as expected with 75% of workflows ran and 25% in progress and every in progress got cancelled on Trigger #1

Rd4dev committed:
Trigger #3 Triggering CI builds #2 worked as expected with 20% of workflows ran and 80% in progress and every in progress got cancelled on Trigger #2

Rd4dev commented:
@adhiamboperes @BenHenning this is a part of #5343 (but a standalone fix), can you PTAL.

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into workflow_canceller_concurrency

adhiamboperes reviewed (approved):
Thanks @Rd4dev! This looks good to me.

Seperately, do you have insights on why the non-app module workflows take long to cancel?

Rd4dev commented:
> Thanks @Rd4dev! This looks good to me.
> 
> Seperately, do you have insights on why the non-app module workflows take long to cancel?

Thanks @adhiamboperes, 

Reg Non-app module cancellation delays, I haven't really dug into why it takes a lot of time. However, based on observation, it predominantly occurs during the Non-app modules' **Build App** step, which seems to wait for certain tasks to finish before the cancellation trigger is hit or recognized. 

I may try to trigger the cancellation once the Build App step is complete to see if other stages cancel immediately without delay (which I believe is the case based on a previous observation, but I will confirm with another set of runs to identify all causes).

Rd4dev committed:
Trigger Build to observe how Non-app module checks gets cut off

Rd4dev committed:
Triggering at Utility Build stage after Build Apps step is over

Rd4dev committed:
Trigger Build before Build App stage

Triggering build after Build App at Utility test stage had the same behaviour of delaying the cut off and it even started the Data tests stage and stopped there

Rd4dev committed:
Triggering instance to see if always is causing the delay

Rd4dev committed:
Triggering to see if always is causing the delay with Static Checks

Rd4dev commented:
@adhiamboperes, based on the observation it seems that every job step that uses the expression `if: ${{ always() }}` exhibit the delay in their cancellation.

Since, the expression [`always`](https://docs.github.com/en/actions/learn-github-actions/expressions#always) 
> Causes the step to always execute, and returns true, even when canceled.

### Example case:
with [Static Checks](https://github.com/oppia/oppia-android/blob/develop/.github/workflows/static_checks.yml) workflow we have:

```yml
jobs:
  linters:
    name: Lint Tests
    runs-on: ubuntu-20.04
    steps:
      - uses: actions/checkout@v2

      .............

      - name: Java lint check
        run: |
          bash /home/runner/work/oppia-android/oppia-android/scripts/checkstyle_lint_check.sh $HOME

      - name: Kotlin lint check
        run: |
          bash /home/runner/work/oppia-android/oppia-android/scripts/ktlint_lint_check.sh $HOME

      - name: Feature flag checks
        run: |
          bash /home/runner/work/oppia-android/oppia-android/scripts/feature_flags_check.sh $HOME

      .............

  script_checks:
    name: Script Checks
    runs-on: ubuntu-20.04
    env:
      CACHE_DIRECTORY: ~/.bazel_cache
    steps:
      - uses: actions/checkout@v2

      .............

      - name: Regex Patterns Validation Check
        if: always()
        run: |
          bazel run //scripts:regex_pattern_validation_check -- $(pwd)

      - name: XML Syntax Validation Check
        if: always()
        run: |
          bazel run //scripts:xml_syntax_check -- $(pwd)

      - name: Testfile Presence Check
        if: always()
        run: |
          bazel run //scripts:test_file_check -- $(pwd)

      - name: Accessibility label Check
        if: always()
        run: |
          bazel run //scripts:accessibility_label_check -- $(pwd) scripts/assets/accessibility_label_exemptions.pb app/src/main/AndroidManifest.xml

      - name: KDoc Validation Check
        if: always()
        run: |
          bazel run //scripts:kdoc_validity_check -- $(pwd) scripts/assets/kdoc_validity_exemptions.pb

      .............

```

### Screenshot:
[[Stacktrace]](https://github.com/oppia/oppia-android/actions/runs/10130346409/job/28011558544?pr=5466)

![Screenshot (1546)](https://github.com/user-attachments/assets/08fc0a99-142c-4003-afc3-efe918dac188)


The **Lint Tests** steps didn't have the expression `if: ${{ always() }}` causing it to stop immediately when cancelled.

With the **Script Checks** steps having the `if: ${{ always() }}` expression set, makes the following steps to execute even after getting cancelled. And only after a certain time they get hit by the cancel trigger and actually stop their execution.

I will investigate further to see if there could be any workaround to stop them immediately and update them.

---

**Edit:**
### Workaround

Seems like there is a workaround and [as per docs](https://docs.github.com/en/actions/learn-github-actions/expressions#always), it is suggested to use the expression **`if: ${{ !cancelled() }}`**, using this allows it to still run a job or step regardless of its success or failure while allowing it to get cancelled. So, I replaced all the `if: ${{ always() }}` with `if: ${{ !cancelled() }}` and also confirmed that this doesn't interrupt the job still being called even if it failed (with the unit test workflow).

> Warning: Avoid using always for any task that could suffer from a critical failure, for example: getting sources, otherwise the workflow may hang until it times out. If you want to run a job or step regardless of its success or failure, use the recommended alternative: if: ${{ !cancelled() }}

[[Related Discussion]](https://github.com/orgs/community/discussions/26303#discussion-4252664)

![Screenshot (1552)](https://github.com/user-attachments/assets/f6ca4797-eb0c-4b95-9785-606e83554c91)

Thanks for highlighting the issue, with this now the workflows get cancelled immediately when a new run is triggered.

![Screenshot (1558)](https://github.com/user-attachments/assets/fe8cc93d-bd9f-4657-a546-b0469cbdd0d8)

Rd4dev committed:
Replacing always with !cancelled to not get affected by success and failure cases but still consider cancellations

Rd4dev committed:
Triggering immediately after replacing always with !cancelled to see if they are getting properly cancelled

Rd4dev committed:
Adding !cancelled to unit tests Check Results job while having a forced failure for TestFileCheck

This should fail the test and eventually call the Check Result job inspite of the failure. If it does then the !cancelled is working as intended replacing always

Rd4dev committed:
Triggering after the TestFileCheck failure run

Rd4dev committed:
Replacing all always expressions with !cancelled since the unit test Check Results calls did get triggered even after failure cases

Rd4dev committed:
Triggering immediately after replacing every always expression to confirm the Non-app module steps too does work in the way it is desired to be

Rd4dev committed:
Triggering to confirm every job is getting cancelled immediately without delay

BenHenning commented on .github/workflows/static_checks.yml:
```diff
@@ -151,39 +155,39 @@ jobs:
 
       - name: Regex Patterns Validation Check
-        if: always()
+        if: ${{ !cancelled() }}
```
We should probably have a comment for each of these lines explaining why it's needed (that it's basically a cancellation-compliant alternative to "if always()" per your findings).

Rd4dev replied:
Sure, added the comment - 
```
# The expression if: ${{ !cancelled() }} runs a job or step regardless of its success or failure while responding to cancellations, 
serving as a cancellation-compliant alternative to if: ${{ always() }} in concurrent workflows.
```
for most of the occurrences, while updating the comments that existed as:
```diff
-if: ${{ always() }} # IMPORTANT: Upload reports regardless of status
+if: ${{ !cancelled() }} # IMPORTANT: Upload reports regardless of success or failure status
```

Would these phrasing be clear enough or would it require additional reference to any discussions?

BenHenning replied:
These seem really good--thanks!

BenHenning reviewed (changes requested):
This is really cool to see @Rd4dev and it looks much more robust than using an external canceler.

I had one in-line comment, but I'm also curious about one thing: per your PR description, it doesn't actually seem like https://github.com/styfle/cancel-workflow-action is being used or am I missing something? It seems that we're relying on GitHub Actions itself to handle cancellation here.

Rd4dev commented:
>  but I'm also curious about one thing: per your PR description, it doesn't actually seem like https://github.com/styfle/cancel-workflow-action is being used or am I missing something?

Uhm.. Sorry if the description was unclear. And I am not sure which one was mentioned here.

If that was the PR description mentioning - "The existing workflow canceller utilized the [cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) by syfle."

By that I was mentioning that the currently existing [workflow_canceller.yml](https://github.com/oppia/oppia-android/blob/develop/.github/workflows/workflow_canceller.yml#L20) workflow file is using the [styfle/cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) (to just reference the existing cancellation method) which is now being replaced with GitHub's concurrency. Am I pointing the same or have I mentioned it incorrectly elsewhere?

Rd4dev committed:
Added in-line comments for explaining the introduction of if: {{ !cancelled() }}

Rd4dev committed:
Added missed out period at the end of the in line comments

Rd4dev committed:
Triggering since a Run app module job was failing

Rd4dev committed:
Triggering once again inbetween the previous check to confirm they get cancelled in progress properly

Rd4dev commented:
Thanks for the review, @BenHenning. I've added the suggested in-line comments to each instance, can you PTAL.

BenHenning reviewed (approved):
Thanks @Rd4dev! This LGTM. Super great to see this fixed!

> > but I'm also curious about one thing: per your PR description, it doesn't actually seem like https://github.com/styfle/cancel-workflow-action is being used or am I missing something?
> 
> Uhm.. Sorry if the description was unclear. And I am not sure which one was mentioned here.
> 
> If that was the PR description mentioning - "The existing workflow canceller utilized the [cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) by syfle."
> 
> By that I was mentioning that the currently existing [workflow_canceller.yml](https://github.com/oppia/oppia-android/blob/develop/.github/workflows/workflow_canceller.yml#L20) workflow file is using the [styfle/cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) (to just reference the existing cancellation method) which is now being replaced with GitHub's concurrency. Am I pointing the same or have I mentioned it incorrectly elsewhere?

Ah, I think I simply misunderstood when reading your PR description. Thanks for clarifying!

BenHenning commented:
Going ahead and merging this since it shouldn't have upstream dependencies, and it'd be nice to immediately begin benefiting from the fix. :)
